clippy_lints/src/mem_replace.rs

```diff
@@ -67,8 +67,127 @@ declare_clippy_lint! {
+            // and we already know that the second argument is an
+            // `Option`, we do not need to check the first
+            // argument's type. All that's left is to get
+            // replacee's path.
+            let replaced_path = match args[0].kind {
+                ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mutable, ref replaced) => {
+                    if let ExprKind::Path(QPath::Resolved(None, ref replaced_path)) = replaced.kind {
+                        replaced_path
+                    } else {
+                        return;
+                    }
+                },
+                ExprKind::Path(QPath::Resolved(None, ref replaced_path)) => replaced_path,
+                _ => return,
+            };
+
+            let mut applicability = Applicability::MachineApplicable;
+            span_lint_and_sugg(
+                cx,
+                MEM_REPLACE_OPTION_WITH_NONE,
+                expr.span,
+                "replacing an `Option` with `None`",
+                "consider `Option::take()` instead",
+                format!(
+                    "{}.take()",
+                    snippet_with_applicability(cx, replaced_path.span, "", &mut applicability)
+                ),
+                applicability,
+            );
+        }
+    }
+}
+
+fn check_replace_with_uninit(cx: &LateContext<'_, '_>, expr: &'_ Expr, args: &HirVec<Expr>) {
+    if let ExprKind::Call(ref repl_func, ref repl_args) = args[1].kind {
+        if_chain! {
+            if repl_args.is_empty();
+            if let ExprKind::Path(ref repl_func_qpath) = repl_func.kind;
+            if let Some(repl_def_id) = cx.tables.qpath_res(repl_func_qpath, repl_func.hir_id).opt_def_id();
+            then {
+                if match_def_path(cx, repl_def_id, &paths::MEM_UNINITIALIZED) {
+                    span_help_and_lint(
+                        cx,
+                        MEM_REPLACE_WITH_UNINIT,
+                        expr.span,
+                        "replacing with `mem::uninitialized()`",
+                        "consider using the `take_mut` crate instead",
+                    );
+                } else if match_def_path(cx, repl_def_id, &paths::MEM_ZEROED) &&
+                        !cx.tables.expr_ty(&args[1]).is_primitive() {
+                    span_help_and_lint(
+                        cx,
+                        MEM_REPLACE_WITH_UNINIT,
+                        expr.span,
+                        "replacing with `mem::zeroed()`",
+                        "consider using a default value or the `take_mut` crate instead",
+                    );
+                }
+            }
+        }
+    }
+}
+
+fn check_replace_with_default(cx: &LateContext<'_, '_>, expr: &'_ Expr, args: &HirVec<Expr>) {
+    if let ExprKind::Call(ref repl_func, _) = args[1].kind {
+        if_chain! {
+            if !in_macro(expr.span) && !in_external_macro(cx.tcx.sess, expr.span);
```

Comment:
```suggestion
            if !in_external_macro(cx.tcx.sess, expr.span);
```

Reply:
Hey we could lint inside macros with the change you suggested but the problem is that the suggestion is incorrect as shown below:
```
error: replacing a value of type `T` with `T::default()` is better expressed using `std::mem::take`
  --> $DIR/mem_replace.rs:27:9
   |
LL |         std::mem::replace($s, Default::default())
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `std::mem::take(s)`
...
LL |     take!(s);
   |     --------- in this macro invocation
```

Notice the variable name needs to be `$s` instead of `s`. This is the reason I am a little hesitant to allow it in local macros unless there is a way to get the macro variable name `$s`. Hope that makes sense and let me know if that's not a problem.

Reply:
You're using `snippet_with_applicability`, which has this at the start:

https://github.com/rust-lang/rust-clippy/blob/143ad5e2d10582d2cd1c74d3bee20b14c04af02e/clippy_lints/src/utils/mod.rs#L495-L497

So the suggestion is `MaybeIncorrect` and won't be applied by running `cargo fix --clippy`. You can test this by adding `// rustfix-only-machine-applicable` to the test. I think if you have to apply the suggestion by hand, you'll write `take($s)` instead of the suggested. 

Another thing you can do is using `span_lint_and_then` and only add the suggestion if the span is not in a macro.

Reply:
Adding the `// rustfix-only-machine-applicable` still causes `cargo fix` to fix the code within the macro. I am pushing the change I have locally. Also I tried the `span_lint_and_then` suggestion but looks like the build fails since there is no fix being applied for internal macros in that case.

I had something like so when using `span_lint_and_then`:
```
span_lint_and_then(
                    cx,
                    MEM_REPLACE_WITH_DEFAULT,
                    expr_span,
                    "replacing a value of type `T` with `T::default()` is better expressed using `std::mem::take`",
                    |db| {
                        if !in_macro(expr_span) {
                            let suggestion = format!("std::mem::take({})", snippet(cx, dest.span, ""));

                            db.span_suggestion(
                                expr_span,
                                "consider using",
                                suggestion,
                                Applicability::MachineApplicable
                            );
                        }
                    }
                );
```

Reply:
Then I would suggest to take the `span_lint_and_then` approach. For fixing the tests you can split up the tests in `mem_replace.rs` and `mem_replace_macro.rs`, where only on `mem_replace.rs` `// run-rustfix` is added.

Reply:
Fixed in [`5d8366`](https://github.com/rust-lang/rust-clippy/pull/4881/commits/5d8366f5fba0727c3f1b1efd4f73cfc9db8f2b3f). Thanks!